### ISSUE TITLE
[QUESTION] find the objects based on value in array of sub documents

### ISSUE THREAD

**Issue by rshillington** [litedb-org/LiteDB#2372]
Consider these simple documents in a collection:
```json
{
  "name":"bob","cards":[{"active":true,"number":1},"active":true,"number":2]
}
{
 "name":"jane","cards":[{"active":true,"number":4},"active":true,"number":5]
}
```
I would try and find the person with card number 4 as
```csharp
var person = collection.Find("$.cards[@.number=4]").FirstOrDefault();
```
but that throws
`LiteDB.LiteException: Expression '$.cards[@.number=4]' are not supported as predicate expression.`

**Comment by mbdavid** [comment]
Hi @rshillington ,

If you need filter based on sub items, you can use:

```SQL
INSERT INTO person VALUES
{
  "name":"bob","cards":[{"active":true,"number":1}, {"active":true,"number":2}]
},
{
 "name":"jane","cards":[{"active":true,"number":4}, {"active":true,"number":5}]
}

SELECT $ FROM person WHERE $.cards[*].number ANY = 4

-- You can create an index to query using this index
CREATE INDEX idx_1 ON person ($.cards[*].number)
```

**Comment by rshillington** [comment]
@mbdavid What I was struggling with is the expression within C#.  Also if I were to try:
```
SELECT $ FROM person WHERE $.cards[*].number ANY = 4 and $.cards[*].active ANY  = true
```
would that select the person if there was a card object with active:true and number:4 --- or would it select a person that has any card with a number of 4 and any card with an active of `true`.

Or to put it another way what would be the functional equivilant to `$elemMatch` in MongoDB?

**Comment by rshillington** [comment]
@mbdavid as per my previous comment I'm now quite certain that if a person has a card object in the cards array with a number of 4 and some other card in the cards array with an active property of true that row will be person object will be found.
In mogoDB speak, what I'm looking for is an `$elemMatch`

**Comment by JKamsker** [comment]
You're right: `$.cards[*].number ANY = 4 AND $.cards[*].active ANY = true` checks the two conditions independently, so it also matches a person whose card 4 is inactive while some other card is active. For `$elemMatch` semantics, filter the array per element and check that something is left:

```csharp
var people = collection.Find("COUNT($.cards[@.number = 4 AND @.active = true]) > 0");

// parameterized
var people2 = collection.Find(BsonExpression.Create(
    "COUNT($.cards[@.number = @0 AND @.active = true]) > 0", 4));
```

`$.cards[@.number=4]` on its own evaluates to the filtered *array*, not a boolean. That's why you get the "not supported as predicate expression" error.

With typed classes and LINQ, use `Where(...).Any()` on the list. `Any(c => ...)` with a condition isn't translated yet; see #2475.

```csharp
col.EnsureIndex(p => p.Cards.Select(c => c.Number));

var result = col.Query()
    .Where(p => p.Cards.Select(c => c.Number).Any(n => n == 4))        // index prefilter
    .Where(p => p.Cards.Where(c => c.Number == 4 && c.Active).Any())   // same-element check
    .ToList();
```

The element filter can't use an index by itself, so on larger collections pair it with a multikey index and an `ANY` prefilter as above. In SQL that is `WHERE $.cards[*].number ANY = 4 AND COUNT($.cards[@.number = 4 AND @.active = true]) > 0`. In LINQ, keep the two as separate `Where` calls: combined with `&&` in one lambda, the optimizer currently doesn't pick the index. I checked this on 5.0.21 and the 6.x prerelease.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
